nexus/src/app/vpc.rs

```diff
@@ -315,18 +329,425 @@ impl super::Nexus {
+            for target in rule.targets.iter_mut() {
+                match target.0 {
+                    external::VpcFirewallRuleTarget::Vpc(ref mut name)
+                        if name.as_str() == "default" =>
+                    {
+                        *name = vpc_name.clone().into()
+                    }
+                    _ => {
+                        return Err(external::Error::internal_error(
+                            "unexpected target in default firewall rule",
+                        ))
+                    }
+                }
+                if let Some(ref mut filter_hosts) = rule.filter_hosts {
+                    for host in filter_hosts.iter_mut() {
+                        match host.0 {
+                            external::VpcFirewallRuleHostFilter::Vpc(
+                                ref mut name,
+                            ) if name.as_str() == "default" => {
+                                *name = vpc_name.clone().into()
+                            }
+                            _ => return Err(external::Error::internal_error(
+                                "unexpected host filter in default firewall rule"
+                            )),
+                        }
+                    }
+                }
+            }
+        }
+        debug!(self.log, "default firewall rules for vpc {}", vpc_name; "rules" => ?&rules);
+        Ok(rules)
+    }
+
+    async fn send_sled_agents_firewall_rules(
+        &self,
+        opctx: &OpContext,
+        vpc: &db::model::Vpc,
+        rules: &[db::model::VpcFirewallRule],
+    ) -> Result<(), Error> {
+        let rules_for_sled = self
+            .resolve_firewall_rules_for_sled_agent(opctx, &vpc, rules)
+            .await?;
+        debug!(self.log, "resolved {} rules for sleds", rules_for_sled.len());
+        let sled_rules_request =
+            sled_agent_client::types::VpcFirewallRulesEnsureBody {
+                rules: rules_for_sled,
+            };
+
+        let vpc_to_sleds =
+            self.db_datastore.vpc_resolve_to_sleds(vpc.id()).await?;
+        debug!(self.log, "resolved sleds for vpc {}", vpc.name(); "vpc_to_sled" => ?vpc_to_sleds);
+
+        let mut sled_requests = Vec::with_capacity(vpc_to_sleds.len());
+        for sled in &vpc_to_sleds {
+            let sled_id = sled.id();
+            let vpc_id = vpc.id();
+            let sled_rules_request = sled_rules_request.clone();
+            sled_requests.push(async move {
+                self.sled_client(&sled_id)
+                    .await?
+                    .vpc_firewall_rules_put(&vpc_id, &sled_rules_request)
+                    .await
+                    .map_err(|e| Error::internal_error(&e.to_string()))
+            });
+        }
+
+        debug!(self.log, "sending firewall rules to sled agents");
+        let results = join_all(sled_requests).await;
+        // TODO-correctness: handle more than one failure in the sled-agent requests
```

Comment:
Is there an issue to track this? I think it's a pretty big and multifaceted work-stream, since it's a potentially long-running, retry-style operation.